### PR TITLE
Update Cookie Manager library to make max cookie number configurable

### PR DESCRIPTION
**Pull request by bartoszherba**
# Configurable max cookie number

### Description
Replace max_cookie_number constant with a value configurable from admin panel.
Default value is set to 50.

### Manual testing scenarios
1. To change maximum cookies number in admin panel go to: Stores -> Configuration -> Web -> Default Cookies Settings -> Cookie Max Number and set required value

2. Use following script in console to bulk create cookies:
for (let i = 0; i < 50; i++) {
 	let t = Math.random().toString(36).substring(7); 
        document.cookie = t+"="+t;   
}

3. Test if exception has been thrown.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=10516) <br/>All committers have signed the CLA.

**Comment by orlangur** [comment]
> At any time, the user agent MAY "remove excess cookies" from the
   cookie store if the number of cookies sharing a domain field exceeds
   some implementation-defined upper bound (such as 50 cookies).

 > At any time, the user agent MAY "remove excess cookies" from the
   cookie store if the cookie store exceeds some predetermined upper
   bound (such as 3000 cookies).

Why did you remove reference to RFC and for which practical purpose do you need more than 50 cookies?

**Comment by bartoszherba** [comment]
@orlangur 
1. I removed rfc by mistake while refactoring consts docblocks according to M2 code style guide. My mistake.

2. I had few projects with live chats and some additional plugins/modules that was using a lot of cookies and therefore I had to increase this limit. Since PhpCookieManager is a Lib class and there is no possible way to extend it and change the max cookie limit I always was overriding it by copy-paste most of the code and set preference. I do not want to mess like that with any lib/core classes so either Manager should be extendable or configurable. According the M2 team to core libs extensibility the latest approach is best.

So, in short, this change is required by projects development practice. Browsers can manage more than 50 cookies and it should be configurable because overriding lib classes yield hard to maintain app.

One of the feature request I found is https://github.com/magento/magento2/issues/7931

**Comment by orlangur** [comment]
Thanks for the reply.

> refactoring consts docblocks according to M2 code style guide

Where is that from? `/**#@+` docblocks are perfectly valid for constant grouping AFAIK.

I have no idea why it was implemented as `PhpCookieManager::MAX_NUM_COOKIES` currently and not `self::MAX_NUM_COOKIES`. I would prefer it as `static::MAX_NUM_COOKIES` rather than Admin UI field so that it can be easily overridden  by inheritance (which is a discouraged approach in M2, but using cookies that much instead of local storage is an outdated approach as well). 

Let's see what other guys think (Cc: @akaplya maybe you remember decisions behind https://github.com/magento/magento2/commit/c945dbe42e5947e29fb977fd44300730e5c46bb7).

**Comment by bartoszherba** [comment]
@orlangur 
regarding const please refer to http://devdocs.magento.com/guides/v2.0/coding-standards/docblock-standard-general.html, the consts section

As for the "static approach", notice that methods are checking that cookie can be set are private and therefore will be run from parent scope. Therefore static will not get overridden const value. And, well, const should be a const, changing its value like this smells like a bad approach for me plus I totally agree with limiting inheritance possibilities which leads to a better architecture but this, on the other hand requires, that classes are following SOLID rules. Cookie manager in current shape is nothing like that. Eventually, we can make max_cookie_number as a constructor's configurable variable (DI approach) but then it still requires a developer to produce at least a module with relevant di.xml
From my point of view in context of projects, customers needs and experience, making max cookie configurable is desired.

**Comment by orlangur** [comment]
Thanks, I somehow missed when http://devdocs.magento.com/guides/v2.2/coding-standards/docblock-standard-general.html#DocBlock-templates was changed in https://github.com/magento/devdocs/commit/7e31ce6a12a5a366440ab1efec731ccd190f54c1.

**Comment by driskell** [comment]
I think we're missing one possible issue here. I've had this occur where a new Magento 2 site went live and the previous site did not care much for the cookie limit. Furthermore, when there are many subdomains active on the domain storing cookies at the domain root it can further increase cookie counts. It does not give a good impression to launch a new site and it simply fail as soon as you change the hostname.

Would be great to get this merged.

**Comment by orlangur** [comment]
@driskell my main concern here about changing `50` to any value (no matter via Admin UI or some other approach) is that I have no idea which is the real limit modern browsers support starting from which cookies will start to be dropped off. Hard limit encourage developers to not bloat the cookies by using local storage or something.

**Comment by driskell** [comment]
@orlangur I can agree with that. I have been thinking about this a little and I wonder if the issue is not the prevention attempt, it's the action taken to prevent.

I am thinking that in any scenario this occurs the problem is not being prevented. If you have 50 cookies already the horse has already bolted.

So I can understand this only as a developer deterrence during development. Which makes me think this should be a developer only mode feature.

Thinking further if we want to prevent issues on Production mode due to limit than maybe not failing is a better option. Instead log some warning maybe even flagging a cache entry to trigger an admin message, "Some users are experiencing too many cookies - please investigate your cookie usage."
It could then possibly have a whitelist of critical cookies to set and refuse setting non-essential ones to try keep under the limit (one would hope the essential session cookie is sent first so always gets a slot.)
I actually think this kind of message would be far better than waking up the ops team at 3am with a sudden burst of 503 errors they can do absolutely nothing about without extensive development work to work out what cookies are on that persons machine and not other peoples (we are finding that due to age of the domain the number of cookies wildly varies machine to machine in our offices alone.)

**Comment by orlangur** [comment]
@driskell this PR is about max cookie number only, does current logic https://github.com/bartoszherba/magento2/blob/885d2e38bc1945556fc25c2ded0521d9238f7da5/lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php#L238 cause 503 in production?

Or do you refer to the maximum cookie size exception? It is not avoidable I believe: https://www.google.com/search?q=maximum+cookie+size

**Comment by driskell** [comment]
Ahhhhh!

https://github.com/magento/magento2/commit/1598c0fa816be99cef010721792755fc7506eced

Perfect! I am using 2.1, great to see it fixes in 2.2 :) Hopefully OP will like this too.

(Just edited - noticed the link wasn't on the magento2 upstream repo - it's definitely in 2.2!)

**Comment by orlangur** [comment]
@driskell nice catch!

@bartoszherba so, "Test if exception has been thrown" does not occur on `develop` in fact, are you satisfied with warning logged?

File: app/code/Magento/Cookie/etc/adminhtml/system.xml [hunk 1]
```diff
@@ -22,14 +22,18 @@
                     <label>Cookie Domain</label>
                     <backend_model>Magento\Cookie\Model\Config\Backend\Domain</backend_model>
                 </field>
-                <field id="cookie_httponly" translate="label" type="select" sortOrder="40" showInDefault="1" showInWebsite="1" showInStore="1">
+                <field id="cookie_max" translate="label" type="text" sortOrder="40" showInDefault="1" showInWebsite="1" showInStore="0">
```
Comment (ishakhsuvarov) [inline app/code/Magento/Cookie/etc/adminhtml/system.xml:25]: This setting is most likely to be set just once, so there is not really a reason to have it in the admin. Consider implementing this as a configuration file.

File: lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php [hunk 2]
```diff
@@ -20,26 +22,32 @@
...
 class PhpCookieManager implements CookieManagerInterface
 {
-    /**#@+
-     * Constants for Cookie manager.
-     * RFC 2109 - Page 15
-     * http://www.ietf.org/rfc/rfc6265.txt
+    /**
+     * Max cookie size
      */
-    const MAX_NUM_COOKIES = 50;
```
Comment (ishakhsuvarov) [inline lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php:32]: Removing a constant is a backwards incompatible change, since some third-party customization could rely on it. Please revert it and add a deprecated annotation.

File: lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php [hunk 3]
```diff
@@ -70,17 +82,20 @@ class PhpCookieManager implements CookieManagerInterface
...
      * @param HttpHeader $httpHeader
+     * @param ScopeConfigInterface $scopeConfig
      */
     public function __construct(
         CookieScopeInterface $scope,
         CookieReaderInterface $reader,
         LoggerInterface $logger = null,
-        HttpHeader $httpHeader = null
+        HttpHeader $httpHeader = null,
+        ScopeConfigInterface $scopeConfig
```
Comment (ishakhsuvarov) [inline lib/internal/Magento/Framework/Stdlib/Cookie/PhpCookieManager.php:92]: Adding a required constructor parameter is a backwards incompatible change, which may affect classes extending this one. Please check code example in [Backwards Compatible Development Guide](http://devdocs.magento.com/guides/v2.1/ext-best-practices/extension-coding/backwards-compatible-development/index.html#adding-a-constructor-parameter) and fix accordingly.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by ishakhsuvarov** [comment]
@bartoszherba Please let us know if you would like to continue with this Pull Requests, since looking at the conversation here it looks like the initial problem is solved.

**Comment by bartoszherba** [comment]
Sorry for delay.
Considering all arguments and changes already made I am happy with the current solution and do not feel like continuing this PR is still required.

**Comment by orlangur** [comment]
@bartoszherba thanks for collaboration and picking up the best solution for practical purposes!

**Comment by ishakhsuvarov** [comment]
@bartoszherba Thanks for collaboration!
